site/src/components/MdxPageContent.tsx

```diff
@@ -22,38 +21,59 @@ export const MdxPageContent: VFC<MdxPageContentProps> = ({
 
   const [headings, setHeadings] = useState<Heading[]>([]);
 
-  useEffect(() => {
-    setHeadings([]);
-  }, [serializedPage]);
-
-  const [currentHeading, setCurrentHeading] = useState<Heading>();
   const [detectHeadingFromScroll, setDetectHeadingFromScroll] =
     useState<boolean>(true);
 
-  const [anchor, setAnchor] = useState<string>(
-    router.asPath.split("#")[1] ?? "",
-  );
+  const currentHeading = useRef<Heading | undefined>(undefined);
 
   useEffect(() => {
-    setAnchor(router.asPath.split("#")[1] ?? "");
-  }, [router]);
+    setHeadings([]);
+
+    return () => {
+      currentHeading.current = undefined;
+      setHeadings([]);
+    };
+  }, [serializedPage]);
+
+  const scrolledOnce = useRef(false);
 
   useEffect(() => {
-    const headingWithCurrentAnchor = headings.find(
-      (heading) => heading.anchor === anchor,
-    );
+    if (headings.length) {
+      const anchor = router.asPath.split("#")[1] ?? "";
+
+      const headingWithCurrentAnchor = headings.find(
+        (heading) => heading.anchor === anchor,
+      );
+
+      const previousRoute = sessionStorage.getItem("previousRoute");
+
+      const shouldScrollToAnchor =
+        // if anchor is empty and we haven't scrolled, prevent it
+        (anchor === "" && scrolledOnce.current) ||
+        // if anchor is not empty, always allow scroll
+        anchor !== "" ||
+        // OR if previous path is either a docs or spec page
+        (previousRoute?.includes("/docs") &&
+          router.asPath?.includes("/docs")) ||
+        (previousRoute?.includes("/spec") && router.asPath?.includes("/spec"));
+
+      if (!scrolledOnce.current) {
+        scrolledOnce.current = true;
+      }
 
-    if (headingWithCurrentAnchor) {
       if (
-        !currentHeading ||
-        headingWithCurrentAnchor.anchor !== currentHeading.anchor
+        headingWithCurrentAnchor &&
+        shouldScrollToAnchor &&
+        (!currentHeading.current ||
+          headingWithCurrentAnchor.element !==
+            currentHeading.current.element) &&
+        document.body.contains(headingWithCurrentAnchor.element)
```

Comment:
This effect was getting stale `headings` data between renders. The condition of `contains` prevents the app from trying to scroll to a stale heading.